### ISSUE TITLE
Feature Request: make it work well when several instances are run concurrently

### ISSUE THREAD

**Issue by expipiplus1** [8go/matrix-commander#31]
```bash
for i in $(seq 1 10); do matrix-commander --message $i & ; done
```

Running the above yields:

- One successful message to the channel
- Two `peewee.OperationalError: database is locked` errors
- Six corrupt messages in the channel `Unable to decrypt: The secure channel with the sender was corrupted`
- One `ERROR: matrix-commander: Message send failed. Sorry.` with no traceback

It would be excellent if multiple instances of this program could run concurrently, or at the very least wait politely for a while for any siblings to finish.

**Comment by 8go** [comment]
Thank you @expipiplus1 for bringing this to my attention. 

Yes, you are right. Concurrency is currently not supported and will cause errors and problems. 

Making the matrix-commander concurrent would be quite some work. 

Just as a workaround, for performance reason it is much better anyway to first prepare all messages and send all N messages with a single `matrix-commander` command. 

Like: 
```
# some preprocessing, preparing all N messages
# calling matrix-commander similar to this:
matrix-commander --message $M1 --message $M2 ... --message $Mn
# this requires less resources and is way faster than calling matrix-commander N times
```

**Comment by expipiplus1** [comment]
Sure, but sometimes that's not possible. For example when MC is being
invoked concurrently from several scripts.

Is it not possible to do something simple like maintain a lock file on any
contended resources?

**Comment by 8go** [comment]
Sure, a "lockfile" or a PID-file is possible, maybe in combination with a flag/argument `--wait` which can be used as:
- if `--wait` is not set and another MC process is running, the newly invoked MC process simply aborts with a message "another MC already running"
- if `--wait` is set  and another MC process is running, the newly invoked MC process will wait (sleep) and periodically (e.g. every 1 or 2 seconds) check if the lockfile (PID file) is still there and by doing so wait until other MC instances have finished, in which case it will itself create the lockfile (PID file) and start running.

From the 2 variants lock-file vs PID-file the PID-file logic is already implemented. 
- See the PID file in e.g. /home/$USER/.run
- See also code: https://github.com/8go/matrix-commander/blob/master/matrix-commander.py#L786-L794
- Also note that multiple MC instance running in parralel are possible if they use different stores, etc.

So, overall what you want is possible. The missing part, logic, is the "check, sleep, recheck" logic. 

It is not high on my priority list. Any nice soul out there that wants to contribute this feature as a PR?

**Comment by 8go** [comment]
Reopened issue as Feature Request in hope to find a contributor for this feature.

**Comment by 8go** [comment]
Reopened issue as Feature Request in hope to find a contributor for this feature.

**Comment by 8go** [comment]
> Thanks for reopening. I wonder if there's some existing program which wraps other programs doing this lock file management. Probably wouldn't take too long to do in bash. Next time I bump into it I might take a look.

Yes, it could also be implemented OUTSIDE of matrix-commander as you point out correctly. You are right, a bash script wrapper could look for the presence of a PID file in the PID directory (e.g. /home/$USER/.run) and "pause" until no more PID files are in the directory.

Only a 99.9% solution because this would still allow 2 MC instances to start concurrently, both check at the same time, both find the dir empty of PID files, both start, .... and then possibly collide. A race condition.

**Comment by expipiplus1** [comment]
well, that particular bug is easy to work around, after writing the pid
file, only start if you read your own pid from that file. But overall
you're right, concurrent programming is tricky :)

**Comment by nirgal** [comment]
@expipiplus1  As a workaround, you could wrap your calls with flock.
I mean rather than run "matrix-commander args" run instead "flock -c mylockfile matrix-commander args"...
This will delay the execution until the lock is available

**Comment by expipiplus1** [comment]
yeah, that's a pretty good idea. thanks

**Comment by xurizaemon** [comment]
Another tactic would be to run parallel instances of `matrix-commander` in containers, eg via docker `matrixcommander/matrix-commander`. Containers could be given whatever shared state is required via a shared mount, and at the same time isolated from interfering with each others running.

**Comment by 8go** [comment]
Thank you all @nirgal @expipiplus1 @xurizaemon for adding your excellent ideas. Please keep sharing. So, others can learn and take some clues, or even implement some of your suggestions. 

For other people: Keep it going and please keep adding more suggestions if you have any! They are welcome! Thanks.

**Comment by 8go** [comment]
See https://github.com/8go/matrix-commander-rs

An idea to improve performance :smile: 

Maybe Rust also allows for more concurrency.

Talk about it to your friends, post about it, spread the word and provide PRs.  :pray:

And give it a :star: if you like the idea.

**Comment by edwinsage** [comment]
If running concurrently is a goal (for example to speed up making numerous requests), you should also be able to achieve it by creating multiple credentials/stores and having each process use a different one, as described here:
https://github.com/8go/matrix-commander/issues/86#issuecomment-1198493742

**Comment by 8go** [comment]
> If running concurrently is a goal (for example to speed up making numerous requests), you should also be able to achieve it by creating multiple credentials/stores and having each process use a different one, as described here:
> [#86 (comment)](https://github.com/8go/matrix-commander/issues/86#issuecomment-1198493742)

Correct. Should be working today.

**Comment by Benjamin-Loison** [comment]
Thanks to @nirgal [comment](https://github.com/8go/matrix-commander/issues/31#issuecomment-909313336).

I personally made sure to have a single `matrix-commander` installed with:
```bash
sudo find / -type f -name 'matrix-commander' 2> /dev/null
```
Then I renamed the given `matrix-commander` to `matrix-commander_without_flock` and I've added the following `matrix-commander` with `+x` permission:
```bash
#!/bin/bash

flock /var/lock/matrix-commander matrix-commander_without_flock "$@"
```
That way I don't modify my scripts but only the `matrix-commander` program.
After this change @expipiplus1 [command line](https://github.com/8go/matrix-commander/issues/31#issue-876287267) works fine.

**Comment by 8go** [comment]
Thank you @Benjamin-Loison for sharing this with everyone.

**Comment by jotwg** [comment]
I'm using 'task spooler' (https://manpages.ubuntu.com/manpages/xenial/man1/tsp.1.html) to queue matrix-commander invocations. Works pretty good for sending notifications from my icinga2 monitoring system via matrix.
